core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantLimit.java

```diff
@@ -45,6 +46,12 @@ public Result apply(LimitNode limit, Captures captures, Context context)
         if (limit.getCount() == 0) {
             return Result.ofPlanNode(new ValuesNode(limit.getId(), limit.getOutputSymbols(), ImmutableList.of()));
         }
+        if (limit.isOrderSensitive()) {
+            if (isScalar(limit.getSource(), context.getLookup())) {
+                return Result.ofPlanNode(limit.getSource());
+            }
+            return Result.empty();
```

Comment:
Why return here? Even if it's order-sensitive, if the number of rows in the input is guaranteed to be lower than the limit, we can remove the limit. This is handled by the `if` block below.

Reply:
removed all the special handling for order sensitive limit in this rule